Item class and overall refactor

kevr opened the pull request:
There's quite a lot involved in this PR. I'll attempt to keep this description as direct as possible while still providing as much information as I can about the current state.

I'll lay this out in the following format, headed by a label of the title of the section:

1. Philosophical Changes
2. Improvements
3. Unfinished

## Philosphical Changes

There are three major changes to the ideas of how we handle some things in the project with this. First, it is replacing the dictionary-based item information structure with a new `Item` class and it's subclasses. Second, it is attempting to vastly reduce the amount of regex used when parsing item text. Last but not least, it is replacing the JSON query generation by allowing `Item` subclasses to specify what their JSON should be.

#### Item Class

This change introduces the `Item` class. The `Item` class is also subclassed into more specific classes, for example: `Ring`, `BodyArmor`, `Weapon`, `Currency`, `Card`, etc. There are also new subclasses which become bases for more specific subclasses, for example: `Searchable`, `Exchangeable`, `Wearable`.

Let's take one prime example. A `Gloves` object is a subclass of `Armor`, which is a subclass of a `Wearable`, which is a subclass of a `Searchable`, which is a subclass of `Item`.

Doing it this way allows us to specify sanitization and JSON generation each step of the way. An `Item` can generate a basic root JSON query structure with basic fields needed for most `Item`s.  Then, a `Wearable` can extend that root JSON query with more details, like item level, all the way up to the leaf subclass where the job is done. The same ideas applies to sanitization of modifiers; that is, an `Armor` subclass sanitizes modifiers differently than a `Weapon` or an `Accessory` does.

Additionally, subclasses may also introduce more attributes that are then used in members that it overrides or further subclasses. The details are up to the subclasses, and not the users of them.

Furthermore, by providing a subclass, we can now check what type of item it is without inspecting strings contained within it. If something is `Exchangeable`, we can check it via `isinstance(item, Exchangeable)`.

#### Item Flow

```
item = parse_item_info(text)
item = item.deduce_specific_object()
item.sanitize_modifiers()
json = item.get_json()

if isinstance(item, Searchable):
  json = create_pseudo_mods(json)
  json = relax_modifiers(json)

query_url = item.query_url(LEAGUE)
...
results = fetch(response_json, isinstance(item, Exchangeable))
...
```

## Improvements

In addition to the `Item` classes, this set of changes also introduces quite a few performance and readability improvements to the codebase. The entire collection of improvements can be found by looking through the messages of the commits involved with this PR, but just to note a few now, below are a list of some of the most promising improvements done.

* Searching item modifiers by text or id is now performed in O(1) time, with the exception of loading them up on the first call.
* `create_pseudo_mods` now identifies mods that it should create pseudo mods out of in amortized O(1) time, by replacing lists with sets and introducing an all-encompassing set of valid identifiers.
* A new `relax_modifiers` function has been introduced, which modifies the `min` and `max` values of all numeric modifiers in a JSON query to be `-/+ 10%` of the original value.
* The use of the `Item` class is much more friendly to those not involved with their implementation, and feels more natural.
* Multiple influenced items are now searchable.
* Every type of item in the game, with the exception of `Beast`s and Synthesis items (most) are searchable.

## Unfinished

* Synthesis mods on Synthesis uniques and bases still need to be finished.
* RestructuredText comments need to be everywhere in the codebase.
* Delve modifiers, Fractured modifiers, other modifiers
* For certain bases, we should lookup the item type instead of the base directly.
* Testing is broken with this PR, needs to be fixed.
* Much more manual extensive testing needs to be done.
* Search results need to be of uniform currency. We should figure out if an item is worth > 1 chaos, or > 1 exalt, or > 1 mirror and reorganize our display to show that currency only. Currently, results come in any currency with the exception of `Exchangeable` queries.
* Drop modifiers that mean absolutely nothing in price lookups; example: `reduced Attribute Requirements`.

This PR will be updated more to finish off some of the `Unfinished` items here. Updates shall be posted, I just wanted to get this in before sleeping so that it was up for some review/looking at. At the time of writing this, the Unfinished and Improvements lists here are not 100% complete. They will also be updated soon.

As far as comments on this PR goes, I would ask that they are currently limited to comments or questions regarding unopinionated details. That is, questions or comments about something that will break, be a bug, or asking about how something works due to confusion/lack of information. There will more than likely be many, many questions and/or comments with this, so I would like to reduce comments regarding things that do not effect the actual usage of the product; example: "Why are we naming this Armor instead of Armour?" We can address further questions or concerns on a more opinionated basis after we've worked out the majority of firm concerns.

sdf010 committed:
Added Item Class

Item class is just a skeleton and we will add methods to it later

sdf010 committed:
Removed unnecessary attributes from Item class

Added function placeholders

sdf010 committed:
Fixed link count

poe will always put the most links at the front so we use .split(" ")[] to get the most links

sdf010 committed:
modified Item class based on discussions in Discord

Item Tier is stored in iLevel, link counting is more robust and influence is now stored as a list

sdf010 committed:
Started on the parsing overhaul in the parse module

sdf010 committed:
Added mirrored attribute to Item class.
Updated parse to work with most items but uniques

sdf010 committed:
Move flask,map,fragment ect section into item section

sdf010 committed:
remove gem and currency to main body, returning items early will prevent the rest of the item code from running so we can do this. this remove the outer if statement

sdf010 committed:
Added support for flavoured uniques

sdf010 committed:
Corrected bugs

sdf010 committed:
edited trade.py to not call POE API every time we get all the mods

sdf010 committed:
added dicts to trade.py so you can search up item mods faster

sdf010 committed:
merging from upstream dev

sdf010 committed:
Item.py now takes a mod and value pair.
We now use the dict in trade.py to search the correct type of mod we want

sdf010 committed:
We can now get all mods for an item and add them to the modifiers list, which is a tuple of ItemModifier and its values.

sdf010 committed:
removing enumerator and mod_index as it is no longer needed

kevr committed:
Improve many things for our item parsing

* Introduce more specific Item subclasses, so we can more easily
  produce json for specific cases, like a Map or a Fragment, without
  as many if/elif branches required.
* Improve get_item_modifiers_by_* functions in utils/trade.
  These now fetch ItemModifier objects directly instead of returning
  the entire dict containing them and their keys.
* Include map iiq/iir/pack_size when parsing an item. Currently,
  use a dict to decide if we have an "other_type" to return.
* Now support map influences.
* Support all item influences by putting them into the Item influence
  list.
* Additionally, replaced logging in the file's affected by the
  `logging` module. We can now enable DEBUG by passing -d or --debug
  to parse.py when starting it. Example: `sudo python3 parse.py --debug`

Caveats: See the 'reduced' to 'increased' conversion in the explicit
parsing section of this commit. We may need to do things like this for
other mods, though I have not yet encountered them as of this commit.

There are also some augmented mods we're doing nothing with, but I'm
not sure that's much of an issue.

Signed-off-by: Kevin Morris <kevr.gtalk@gmail.com>

kevr committed:
Properly close colors when logging

This was done in the previous commit. This is a text-only
commit detailing that this was done.

We should always immediately close colors when logging via
`Fore.RESET`, and avoid spaces being colored differently.

Additionally, Fore.RESET is better than using Fore.WHITE, as
it's more standard and will not force white text when the user's
terminal settings are different.

Signed-off-by: Kevin Morris <kevr.gtalk@gmail.com>

kevr committed:
Introduce the ability to deduce items into their specific objects

We can now parse an item, which returns an Item type, then call
deduce_specific_object on that Item to convert it into it's
real subclass in O(n) where n = len(NINJA_BASES) time on first call,
and O(1) time on subsequent calls.

On my laptop, first call takes roughly 8.8 milliseconds to run, and
subsequent calls take 0.000001 seconds to run; a fraction of a
millisecond.

Signed-off-by: Kevin Morris <kevr.gtalk@gmail.com>

kevr committed:
Remove timeit import

Signed-off-by: Kevin Morris <kevr.gtalk@gmail.com>

kevr committed:
More progress on parse rework + new lookup + Item classes

With this commit, we can now lookup items with the old pricing
algorithm from the results we get + GUI generation.

Notable improvements:
    * Metamorph organs can now be priced
    * All item classes (with the exception of Beast) that exist can be
      looked up and priced.

Still missing:
    * parsing out Delve, Veiled, and other modifiers (see what we don't
      yet handle at the bottom of parse_item_info).
    * Overall cleanup; removing things we don't need anymore out of
      `parse.py`.
    * Pricing based on a uniform currency -- We currently sort through
      any results we find, regardless of currency, and make averages
      based on the numbers they give. We need to be converting currency
      from results we find to a uniform currency (chaos, exalted).
      This is something that is still an issue, afaik, in the current
      master/dev algorithm. We should improve this before the PR is
      actually created toward upstream.
    * Drop modifiers that are unimportant; example: reduced Attribute
      Requirements.
    * Relax the values that we use in the actual lookup json to produce
      more results. Currently, on items like 6L Loreweave, we match up mods
      so closely that we really do not get many results. This will
      require some serious consideration.
    * Quite a few more things... do note, that this is meant to be a
      partial update and is not meant to be something that is
      "finished."
    * Some more RestructuredText comments for classes and methods.

Signed-off-by: Kevin Morris <kevr.gtalk@gmail.com>

kevr committed:
Moved around the locations of a lot of our `parse` functions

* Began work on synthesis uniques/bases. Conversion works, but
  some synthesis-item specific mods are currently not sanitized.

Signed-off-by: Kevin Morris <kevr.gtalk@gmail.com>

kevr committed:
Item classes are now used throughout the codebase

Modifications:
    * Comment improvements throughout the codebase
    * Organ json query status fixed to be online only
    * Completed item subclassing; added Essence, Scarab, Vial. All
      currency rarity items are handled by the existing Currency subclass
    * Introduced `utils.mods.relax_modifiers` method, which similar to
      create_pseudo_mods, looks at modifiers in a json query and modifies
      their min/max values to be -/+ 10% of the original value
    * Only create pseudo mods and relax modifiers when searching if the
      item is a Wearable
    * Raise `NotFoundException` when we receive no results
    * search_ninja_base now uses our new Item classes passed to it
    * open_trade_site now uses our new Item classes passed to it
    * wiki_lookup now uses our new Item classes passed to it
    * Print exception trace when one was unexpectedly raised during
      hotkey press
    * Reorganization of some variables to more sensible locations
    * Renaming of some variables to follow PEP-8
    * Removed 'max' value specifications in create_pseudo_mods
    * Synthesis uniques and bases are now converted to their matching
      Item subclasses

Still needed:
    - Tests are broken; they need to be fixed before the PR is complete
    - Progress on Synthesis item mods have been started, however, they
      are currently broken and need to be completely finished
    - More thorough testing on items throughout the trade website to
      identify any unforeseen issues

This is committed currently to update the current PR state to what's
most recent. However, there are still some things here that need to be
worked out and tested more thoroughly. The bulk of the work is done, and
our Item class infrastructure is working.

Hurray!

Signed-off-by: Kevin Morris <kevr.gtalk@gmail.com>

rhzk commented on utils/mods.py:
```diff
@@ -0,0 +1,186 @@
+                total_ele_resists += int(i["value"]["min"])
+                combined_filters.append(i)
+
+            # Maximum life
+            elif i["id"] in life_ids:
+                total_life += int(i["value"]["min"])
+                combined_filters.append(i)
+
```
Can you add a elif check for strength as well and add strength/10 to total_life

kevr replied:
In a later PR, yes. However, the primary goal with this PR are the Item class/parsing/searching changes.

It has been waiting for some time to get in so that we could have other developers contributing to the product containing these changes, so our main concern at this point is getting the new flow with these original goals in.

rhzk replied:
Sure, just adding a note so it wont be forgotten

kevr committed:
Add Jewel subclass

Signed-off-by: Kevin Morris <kevr.gtalk@gmail.com>

kevr committed:
Magic Wearable's matching

With get_magic_type, we can now extract both the base and the
item type of any magic item. We now use this to deduce self.base
and subclass type of a magic wearable item.

Signed-off-by: Kevin Morris <kevr.gtalk@gmail.com>

kevr committed:
Support magic map lookups

We now query and store maps from
https://www.pathofexile.com/api/trade/data/items, which provides a
complete listing of maps by type (or base). We can now extend this
to look up for other things when needed.

Signed-off-by: Kevin Morris <kevr.gtalk@gmail.com>

kevr committed:
Add mock data for all item metadata fetch queries

mockBases: poe.ninja bases
mockItems: Path of Exile items
mockModifiers: Path of Exile item modifiers

Signed-off-by: Kevin Morris <kevr.gtalk@gmail.com>

kevr commented:
This PR should be taken over by sdf or somebody else.
